TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/ContentCache.php

```diff
@@ -177,7 +177,7 @@ public function createDynamicCachedSegment($content, $typoScriptPath, array $con
             'context' => $this->serializeContext($contextVariables),
         ];
 
-        return self::CACHE_SEGMENT_START_TOKEN . $this->randomCacheMarker . 'evalCached=' . $identifier . self::CACHE_SEGMENT_SEPARATOR_TOKEN . $this->randomCacheMarker . json_encode($segmentData) . self::CACHE_SEGMENT_SEPARATOR_TOKEN . $this->randomCacheMarker . $content . self::CACHE_SEGMENT_END_TOKEN . $this->randomCacheMarker;
+        return self::CACHE_SEGMENT_START_TOKEN . $this->randomCacheMarker . 'evalCached=' . $identifier . self::CACHE_SEGMENT_SEPARATOR_TOKEN . $this->randomCacheMarker . json_encode($segmentData) . self::CACHE_SEGMENT_SEPARATOR_TOKEN . $this->randomCacheMarker . $this->processCacheSegments($content, false) . self::CACHE_SEGMENT_END_TOKEN . $this->randomCacheMarker;
```

Comment:
Well, this basically removes caching for nested segments with cache mode `cached` (because it strips the markers). I can't say right now if this causes problems because mode `dynamic` basically also is `cached` but with an uncached identifier (discriminator) depending on the context, but I have the  feeling it's not fully correct.

I'd say a fix should rather go in the processing side of the content because segments with `evalCached` seems to not replace the inner markers as we do with cached segments recursively. My best bet: `replaceUncachedPlaceholders`.

Reply:
I could think that this is the issue:

    public function getCachedSegment($uncachedCommandCallback, $typoScriptPath, $cacheIdentifierValues, $addCacheSegmentMarkersToPlaceholders = false)
    {
        $cacheIdentifier = $this->renderContentCacheEntryIdentifier($typoScriptPath, $cacheIdentifierValues);
        $content = $this->cache->get($cacheIdentifier);

        if ($content === false) {
            return false;
        }

        $i = 0;
        do {
            $replaced = $this->replaceCachePlaceholders($content, $addCacheSegmentMarkersToPlaceholders);
            if ($replaced === false) {
                return false;
            }
            if ($i > self::MAXIMUM_NESTING_LEVEL) {
                throw new Exception('Maximum cache segment level reached', 1391873620);
            }
            $i++;
        } while ($replaced > 0);

        $this->replaceUncachedPlaceholders($uncachedCommandCallback, $content);

        ...
    }

Method `replaceCachePlaceholders` is called recursively but `replaceUncachedPlaceholders` is only called once at the end. So a solution might be to call both in the do ... while and check if any of them still replaced a segment and otherwise stop.

Reply:
Thanks for your feedback! I was pretty uncertain about my fix - it was a shot in the dark more or less.
The caching implementation could probably use some refactoring to make it maintainable by mere mortals - or maybe it just *is* that complex ;)